deprecate tokens, use API Keys (#4449)
PrefectHQ/prefect: docs/orchestration/concepts/api_keys.md

# API Keys <Badge text="Cloud"/>

API Keys are how clients authenticate with the Prefect Cloud API.  They encapsulate the identity of a User or a Service Account.  Ultimately, all clients that interact with the Prefect Cloud API must provide an API Key as a Bearer Token included in the request header.

> For more information on how to use these keys to interact with the GraphQL API, go [here](api.html).

Reviewer: ```suggestion
See our [API documentation](api.html) more information on how to use these keys to interact with the GraphQL API.

```

### User API Keys

> Users can generate API Keys to serve as personal access keys.  When creating an API key, you can configure the following attributes:

Reviewer: Introducing the personal access key here is confusing

```suggestion
Users can generate API Keys to interact with the API with their personal permissions.  When creating an API key, you can configure the following attributes:
```

Author: I was avoiding changing that particular language but I agree

- **API Key Name**: The name of this key
- **API Key Expiration**: An optional expiration date for the key - if no expiration is provided, the key will never expire
- **Tenant**: The tenant to associate with the key, by default - this key will inherit its associated user's permissions in all tenants to which the User has an existing membership, but the default tenant will be used when no tenant is specified.  Revoking permissions or memberships of that user will be reflected in the token.

To generate an API key for your User, navigate to User > API Keys within the UI and click "Create an API Key".

::: tip Key Management
Best Practice: When tempted to create a long-lived token for CI, a Prefect Agent, or any use beyond local development, use a Service Account API Key.  Your User API Key is effectively your identity and should be treated as such.
:::

### Service Account API Keys

> Create API Keys restricted to a tenant by creating API Keys associated with Service Accounts in that tenant.  These API Tokens have the same permissions as their Service Account.  To create and manage your tenant's Service Accounts and their associated API keys, navigate to Team > Service Accounts.  

Reviewer: ```suggestion
Users can generate API Keys for Service Accounts, which have permissions restricted to the tenant the Service Account belongs to. These keys will not inherit the full permissions of the user creating them. To create and manage your tenant's Service Accounts and their associated API keys, navigate to Team > Service Accounts.  
```

## Querying for Key Information

> To query for information about API Keys with GraphQL, execute the following query against `https://api.prefect.io`. This will return API Key information, but never the value of the key.  All User API Keys for the active User will be returned, along with any Service Account API Keys which the API Key is privileged to see in the the API Key's default tenant.  Only and all Administrators are privileged to see all Service Accounts--because all Service Accounts have the Administrator role, they can also see all API Keys for any Service Account in the tenant.

Reviewer: ```suggestion
To query for information about API Keys with GraphQL, execute the following query against `https://api.prefect.io`. This will return API Key metadata, but the value of the key cannot be retrieved.
```
This next portion makes sense to me but is _very_ confusing. I'm not sure how to rewrite it.
`All User API Keys for the active user will be returned, along with any Service Account API Keys which the API Key is privileged to see in the the API Key's default tenant.  Only and all Administrators are privileged to see all Service Accounts--because all Service Accounts have the Administrator role, they can also see all API Keys for any Service Account in the tenant.`

Author: Completely agree--but the behaviour is confusing if we don't note SOMETHING here imho.  Leaving this suggestion open

auth_token = SERVICE_ACCOUNT_API_KEY
```

## Deprecation of User Access Tokens and API Tokens

> API Keys replace the deprecated User Access Tokens and API Tokens, which previously followed a different paradigm.  In effect, User API Keys can be used in place of Personal Access Tokens, and Service Account API Keys should replace API Tokens (which can no longer be created in the UI), though they function slightly differently.

Reviewer: ```suggestion
API Keys replace the deprecated User Access Tokens and API Tokens, which used a different authentication paradigm. In effect, User API Keys can be used in place of Personal Access Tokens, and Service Account API Keys should replace API Tokens.
```
